raiden/waiting.py

```diff
@@ -42,6 +54,220 @@
+    raiden: "RaidenService"
+    canonical_identifier: CanonicalIdentifier
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        if channel_state is not None:
+            coop_settle = channel_state.our_state.initiated_coop_settle
+            assert coop_settle is not None, "must be set"
+            return self.raiden.get_block_number() > coop_settle.expiration
+        return False
+
+
+@dataclass
+class ChannelCoopSettleSuccess(ChannelStateCondition):
+    raiden: "RaidenService"
+    canonical_identifier: CanonicalIdentifier
+
+    def __post_init__(self) -> None:
+        assert self.raiden.wal is not None, "must be set"
+        self._stream = self.raiden.wal.storage.get_state_changes_stream(0.1)
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        state_changes = next(self._stream)
+        for state_change in state_changes:
+            if (
+                isinstance(state_change, ContractReceiveChannelSettled)
+                and state_change.canonical_identifier == self.canonical_identifier
+            ):
+                return True
+        return False
+
+
+@dataclass
+class And(ChannelStateCondition):
+    conditions: Iterable[ChannelStateCondition]
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        return all(condition(chain_state, channel_state) for condition in self.conditions)
+
+
+@dataclass
+class Or(ChannelStateCondition):
+    conditions: Iterable[ChannelStateCondition]
+
+    def evaluate(
+        self, chain_state: ChainState, channel_state: Optional[NettingChannelState]
+    ) -> bool:
+        return any(condition(chain_state, channel_state) for condition in self.conditions)
+
+
+@dataclass
+class ChannelStateWaiter:
+    raiden: "RaidenService"
+    retry_timeout: float
+    token_network_registry_address: TokenNetworkRegistryAddress
+    token_address: TokenAddress
+    partner_address: Address
+
+    def _get_channel_state(self, chain_state: ChainState) -> Optional[NettingChannelState]:
+        return _get_channel_state_by_partner_address(
+            chain_state,
+            self.token_network_registry_address,
+            self.token_address,
+            self.partner_address,
+        )
+
+    def wait(self, condition: ChannelStateCondition) -> None:
+        chain_state = views.state_from_raiden(self.raiden)
+        while not condition(chain_state, self._get_channel_state(chain_state)):
+            assert self.raiden.is_running(), ALARM_TASK_ERROR_MSG
+            assert self.raiden.alarm.is_running(), ALARM_TASK_ERROR_MSG
+            log.debug(
+                "Waiting on channel",
+                node=to_checksum_address(self.raiden.address),
+                partner_address=to_checksum_address(self.partner_address),
+                condition=condition,
```

Comment:
I still think this will need some more work, for example optimizing the `Condition`s `__str__` / `__repr__` output (or maybe even adding a human readable description to them).

But this doesn't need to happen in this PR.